src/mam4xx/mo_sethet.hpp

```diff
@@ -0,0 +1,177 @@
+
+//=================================================================================
+KOKKOS_INLINE_FUNCTION
+void calc_precip_rescale(
+    ColumnView cmfdqr,   // dq/dt for convection [kg/kg/s] //in
+    ColumnView nrain,    // stratoform precip [kg/kg/s] //in
+    ColumnView nevapr,   // evaporation [kg/kg/s] // in
+    ColumnView precip) { // precipitation [kg/kg/s] // out
+  // -----------------------------------------------------------------------
+  // calculate precipitation rate at each grid
+  // this is added to rescale the variable precip (which can only be positive)
+  // to the actual vertical integral of positive and negative values.
+  // This removes point storms
+  // -----------------------------------------------------------------------
+
+  Real total_rain; // total rain rate (both pos and neg) in the column
+  Real total_pos;  // total positive rain rate in the column
+
+  total_rain = 0.0;
+  total_pos = 0.0;
+  for (int kk = 0; kk < pver; kk++) {
+    precip(kk) = cmfdqr(kk) + nrain(kk) - nevapr(kk);
+    total_rain = total_rain + precip(kk);
+    if (precip(kk) < 0.0)
+      precip(kk) = 0.0;
+    total_pos = total_pos + precip(kk);
+  }
+
+  if (total_rain <= 0.0) {
+    for (int kk = 0; kk < pver; kk++) {
+      precip(kk) = 0.0; // set all levels to zero
+    }
+  } else {
+    for (int kk = 0; kk < pver; kk++) {
+      precip(kk) = precip(kk) * total_rain / total_pos;
+    }
+  }
+
+} // end subroutine calc_precip_rescale
+
+//=================================================================================
+KOKKOS_INLINE_FUNCTION
+void gas_washout(      // const ThreadTeam &team,
+    int plev,          // calculate from this level below //in
+    Real xkgm,         // mass flux on rain drop //in
+    Real xliq_ik,      // liquid rain water content [gm/m^3] // in
+    ColumnView xhen_i, // henry's law constant
+    ColumnView tfld_i, // temperature [K]
+    ColumnView delz_i, // layer depth about interfaces [cm]  // in
+    ColumnView xgas) { // gas concentration // inout
+  //------------------------------------------------------------------------
+  // calculate gas washout by cloud if not saturated
+  //------------------------------------------------------------------------
+  // FIXME: BAD CONSTANTS
+  Real allca = 0.0; // total of ca between level plev and kk [#/cm3]
+  Real xca, xeqca;
+  Real const0 = boltz_cgs * 1.0e-6; // [atmospheres/deg k/cm^3]
+  Real geo_fac = 6.0; // geometry factor (surf area/volume = geo_fac/diameter)
+  Real xrm = .189;    // mean diameter of rain drop [cm]
+  Real xum = 748.0;   // mean rain drop terminal velocity [cm/s]
+
+  //-----------------------------------------------------------------
+  //       ... calculate the saturation concentration eqca
+  //-----------------------------------------------------------------
+  for (int kk = 0; kk < plev;
+       kk++) { // FIXME: not sure if this should be a Kokkos for or not...
```

Comment:
@odiazib @jeff-cohere @mjs271 looking for input here if this should be a Kokkos for or not?

Reply:
The rule we've been using is this: if all the levels (data related to the index k) are independent, then it's safe to use a Kokkos parallel_for.

Reply:
Okay I put the first part in a parallel_for then the second part in a regular for because it does a check after adding each element to a sum - not sure if that can be done in a parallel_reduce... but closer regardless!